### DISCUSSION TITLE
Compacting the size of the ever-growing WSL images

### DISCUSSION THREAD

**Discussion by okibcn** [microsoft/WSL#9566] [General]
Using virtual drives is a great idea. But, ext4 filesystem optimization inside VHDX images is not so well supported by the optimization tools. NTFS partitions are well optimized, but ext4 partitions are not so well optimized, and WSL suffers from this problem.

Since WSL doesn't provide a good solution for this, I have coded a small utility **[wslcompact](https://github.com/okibcn/wslcompact)** that compacts the WSL partitions. This is the usage of the tool:
```
   Usage: wslcompact [OPTIONS] [DISTROS]

   wslcompact compacts the images of WSL distros by removing unsused space.
   If no option is provided, it will default to info mode, without modifying any image.
   If no distro is provided it will process all the installed images.
   NOTE: WSL will be shutdown for compacting the images.

   Options:
   no opt. Provides name, image file path, current size, and estimated new size information.
       -c   Compacting mode: process the selected distros compacting the images.
       -y   replaces selected images without asking for confirmation.
       -d   Enable the processing of data images. Default is disabled.
       -h   Prints this help

   Examples:
       wslcompact
       wslcompact -c -d
       wslcompact -c -y Ubuntu Kali
```
**[WSLCompact v4](https://github.com/okibcn/wslcompact)** performs the compact process in a 'safe' mode. So, the user can cancel safely the, sometimes long, process without side effects.
The user can decide if the new image is replaced after the result, so the final decision is on the user. Most errors should be tracked now, but do not panic, should you receive an error message, it won't modify any image without the user's approval (unless you use the -y option).

And this is an example of a session:
```
PS> wslcompact -c Ubuntu
 WSL compact, v4.2023.01.31
 (C) 2023 Oscar Lopez
 wslcompact -h for help. For more information visit: https://github.com/okibcn/wslcompact

 Distro's name:  Ubuntu
 Image file:     C:\Users\Oki\WSL\Ubuntu\ext4.vhdx
 Current size:   12864 MB
 Estimated size: 7700 ± 188 MB
 The estimated process time using an SSD is about 2 minutes.
 NOTE: You can safely cancel at any time by pressing Ctrl-C
 Import in progress, this may take a few minutes.
The operation completed successfully.
 New Image compacted from 12864 MB to 7728 MB
 Do you want to apply changes and use the new image (y/N): y
 Image replaced for distro: Ubuntu
```

Any thoughts on improving the tool would be welcome. However, it would be better if the WSL team releases something similar. A good improvement would be an improvement in the export function, not only making a copy of the VHD file but dumping the contents in a newly created VHD, which would considerably reduce the size of the backup.

Any work in this direction?

**Comment by Biswa96** [comment]
Isn't there already Optimize-Vhd command available in Powershell? It can throw away unused data in the ext4.vhdx file.

**Reply by klakiervx** [reply]
Doesn't Optimize-Vhd require Hyper-V to be installed? I don't have Optimize-Vhd (or any other *-Vhd) command available in my PowerShell.

**Reply by Biswa96** [reply]
Probably, not sure though. You can enable it in optional feature in control panel or with dism command.

**Comment by okibcn** [comment]
@Biswa96 yes, there are a couple of ways: using `Optimize-VHD`, and `DISKPART`. However, both require elevated prompt, and it is not efficient. In some cases, they even result in larger images, for instance, if you have run `fstrim /` or 'zerofree' inside some distros. And, in any case, as I stated in the OP, those methods work well when the fs on top is NTFS, but ext4 is not enjoying the same attention.

**Reply by Biswa96** [reply]
I agree with your point. Some clarifications.

Optimize-Vhd itself does **not** need admin permission if the current user has full r/w permission with the ext4.vhdx file.

Also the procedure with Optimize-Vhd does not require any diskpart commands.

**Reply by okibcn** [reply]
Optimize-Vhd doesn't require diskpart, of course, both processes can do more or less the same regarding drive compaction. Diskpart is somehow more convenient as it is readily available in all systems without the need for installing any extra package.

But, in any case, both are far from the potential solution if `wsl --export --vhd` could export optimized vhdx images instead of just copying them, it would be a great step forward.
